File: app/src/main/java/com/artamonov/millionplanets/modules/ModulesInfoActivity.java
package com.artamonov.millionplanets.modules;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;
import com.artamonov.millionplanets.R;
import com.artamonov.millionplanets.model.Module;
import com.artamonov.millionplanets.model.ObjectModel;
import com.artamonov.millionplanets.model.User;
import com.artamonov.millionplanets.utils.Utils;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.WriteBatch;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ModulesInfoActivity extends AppCompatActivity {
    FirebaseFirestore firebaseFirestore = FirebaseFirestore.getInstance();
    User userList = new User();
    ObjectModel objectModelList = new ObjectModel();
    List<Module> moduleList = new ArrayList<>();

    private DocumentReference userDocumentReference;
    private DocumentReference modulesDocumentReference;
    private DocumentReference planetDocumentReference;
    private FirebaseUser firebaseUser;
    private TextView tv_YourShip;
    private TextView tvUserCash;
    private TextView tvUserBank;
    private Button btnBuy;
    private Button btnSell;
    private RecyclerView rvShipyard;
    private List<String> shipsArrayList = new ArrayList<>();
    private String yourShip;
    private Module module;
    private int position;
    private int freeSlots;
    private String currentModule;
    private ConstraintLayout parentLayout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.modules_info);
        Intent intent = getIntent();
        position = intent.getIntExtra("position", -1);

        firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
        userDocumentReference =
                firebaseFirestore.collection("Objects").document(firebaseUser.getDisplayName());
        modulesDocumentReference =
                firebaseFirestore.collection("Modules").document(firebaseUser.getDisplayName());

        parentLayout = findViewById(R.id.modulesinfo_parentLayout);
        TextView tvClass = findViewById(R.id.modulesinfo_class);
        TextView tvCost = findViewById(R.id.modulesinfo_cost);
        TextView tvDamageHP = findViewById(R.id.modulesinfo_damageHp);
        TextView tvDamageShield = findViewById(R.id.modulesinfo_damageShield);
        tvUserCash = findViewById(R.id.modulesinfo_user_cash);
        btnBuy = findViewById(R.id.modulesinfo_buy);
        btnSell = findViewById(R.id.modulesinfo_sell);

        module = Utils.getCurrentModuleInfo(position);
        tvClass.setText(module.getModuleClass());
        tvCost.setText(String.valueOf(module.getPrice()));
        tvDamageHP.setText(String.valueOf(module.getDamageHP()));
        tvCost.setText(String.valueOf(module.getPrice()));
        tvDamageShield.setText(String.valueOf(module.getDamageShield()));
    }

    @Override
    protected void onStart() {
        super.onStart();
        firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
        userDocumentReference =
                firebaseFirestore.collection("Objects").document(firebaseUser.getDisplayName());
        modulesDocumentReference =
                firebaseFirestore.collection("Modules").document(firebaseUser.getDisplayName());
        userDocumentReference.addSnapshotListener(
                this,
                new EventListener<DocumentSnapshot>() {
                    @Override
                    public void onEvent(
                            @javax.annotation.Nullable DocumentSnapshot doc,
                            @javax.annotation.Nullable FirebaseFirestoreException e) {
                        if (doc.exists()) {
                            userList.setMoney(doc.getLong("money").intValue());
                            tvUserCash.setText(Integer.toString(userList.getMoney()));
                        }
                    }
                });
        modulesDocumentReference.addSnapshotListener(
                this,
                new EventListener<DocumentSnapshot>() {
                    @Override
                    public void onEvent(
                            @javax.annotation.Nullable DocumentSnapshot doc,
                            @javax.annotation.Nullable FirebaseFirestoreException e) {
                        if (doc.exists()) {
                            currentModule = doc.getString("weaponName");
                            Log.i("myLogs", "current Module " + currentModule);
                            setOnBuySellButtonVisibility(currentModule);
                        } else {
                            setOnBuySellButtonVisibility("");
                        }
                    }
                });
    }

    private void setOnBuySellButtonVisibility(String currentModule) {
        if (currentModule.equals(module.getName())) {
            btnBuy.setEnabled(false);
            btnBuy.setBackgroundColor(getResources().getColor(R.color.grey));
            btnSell.setEnabled(true);
            btnSell.setBackgroundColor(getResources().getColor(R.color.colorAccent));
        } else {
            btnBuy.setEnabled(true);
            btnBuy.setBackgroundColor(getResources().getColor(R.color.colorAccent));
            btnSell.setEnabled(false);
            btnSell.setBackgroundColor(getResources().getColor(R.color.grey));
        }
    }

    public void onBuy(View view) {

        if (!ifEnoughMoney()) {
            Toast.makeText(this, "Not enough money to buy the module!", Toast.LENGTH_SHORT).show();
            return;
        }

        WriteBatch batch = firebaseFirestore.batch();
        Map<String, Object> moduleMap = new HashMap<>();
        moduleMap.put("damageHP", module.getDamageHP());
        moduleMap.put("damageShield", module.getDamageShield());
        moduleMap.put("weaponClass", module.getModuleClass());
        moduleMap.put("weaponName", module.getName());
        batch.set(modulesDocumentReference, moduleMap);
        batch.update(userDocumentReference, "money", userList.getMoney() - module.getPrice());
        batch.commit()
                .addOnCompleteListener(
                        new OnCompleteListener<Void>() {
                            @Override
                            public void onComplete(@NonNull Task<Void> task) {
                                btnBuy.setEnabled(false);
                                btnBuy.setBackgroundColor(getResources().getColor(R.color.grey));
                                btnSell.setEnabled(true);
                                btnSell.setBackgroundColor(
                                        getResources().getColor(R.color.colorAccent));
                                Snackbar.make(
                                                parentLayout,
                                                "The item was successfully bought",
                                                Snackbar.LENGTH_SHORT)
                                        .show();
                            }
                        });
    }

    private boolean ifEnoughMoney() {
        return userList.getMoney() > module.getPrice();
    }

    public void onGoBackToMainOptions(View view) {
        finish();
    }

    public void onSell(View view) {
        WriteBatch batch = firebaseFirestore.batch();
        Map<String, Object> moduleMap = new HashMap<>();
        moduleMap.put("damageHP", Utils.getCurrentModuleInfo(0).getDamageHP());
        moduleMap.put("damageShield", Utils.getCurrentModuleInfo(0).getDamageShield());
        moduleMap.put("weaponClass", Utils.getCurrentModuleInfo(0).getModuleClass());
        moduleMap.put("weaponName", Utils.getCurrentModuleInfo(0).getName());
        batch.set(modulesDocumentReference, moduleMap);
        batch.update(userDocumentReference, "money", userList.getMoney() + module.getPrice() / 2);
        batch.commit()
                .addOnCompleteListener(
                        new OnCompleteListener<Void>() {
                            @Override
                            public void onComplete(@NonNull Task<Void> task) {
                                btnBuy.setEnabled(true);
                                btnBuy.setBackgroundColor(
                                        getResources().getColor(R.color.colorAccent));
                                btnSell.setEnabled(false);
                                btnSell.setBackgroundColor(getResources().getColor(R.color.grey));
                                Snackbar.make(
                                                parentLayout,
                                                "The item was successfully sold",
                                                Snackbar.LENGTH_SHORT)
                                        .show();
                            }
                        });
    }
}

File: app/src/main/java/com/artamonov/millionplanets/modules/ModulesInfoActivity.kt
package com.artamonov.millionplanets.modules

import android.os.Bundle
import android.view.View
import androidx.core.content.ContextCompat
import com.artamonov.millionplanets.R
import com.artamonov.millionplanets.base.BaseActivity
import com.artamonov.millionplanets.databinding.ActivityModulesInfoBinding
import com.artamonov.millionplanets.model.Module
import com.artamonov.millionplanets.model.User
import com.artamonov.millionplanets.model.Weapon
import com.artamonov.millionplanets.utils.getCurrentModuleInfo
import com.artamonov.millionplanets.utils.getCurrentShipInfo
import com.artamonov.millionplanets.utils.showSnackbarError

class ModulesInfoActivity : BaseActivity() {
    private var userList = User()
    private var module: Module? = null
    private var position: Int = 0

    lateinit var binding: ActivityModulesInfoBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityModulesInfoBinding.inflate(layoutInflater)
        setContentView(binding.root)
        val intent = intent
        position = intent.getIntExtra("position", -1)
        module = position.getCurrentModuleInfo()
        binding.modulesinfoClass.text = module?.moduleClass
        binding.modulesinfoCost.text = module?.price.toString()
        binding.modulesinfoDamageHp.text = module?.damageHP.toString()
        binding.modulesinfoCost.text = module?.price.toString()

        binding.modulesinfoBuy.setOnClickListener {

            if (!isEnoughMoneyToBuy()) {
                showSnackbarError(getString(R.string.modules_info_not_enough_money))
                return@setOnClickListener
            }

            val batch = firebaseFirestore.batch()
            userList.weapon?.add(Weapon(position.toLong(), isSlotAvailable()))
            batch.update(userDocument, "weapon", userList.weapon)
            batch.update(userDocument, "money", userList.money!! - module?.price!!)
            batch.commit()
                    .addOnCompleteListener {
                        binding.modulesinfoSell.isEnabled = true
                        binding.modulesinfoSell.setBackgroundColor(
                                ContextCompat.getColor(this, R.color.colorAccent))
                        showSnackbarError(getString(R.string.modules_info_item_was_bought_successfully))
                    }
        }

        binding.modulesinfoSell.setOnClickListener {
            val batch = firebaseFirestore.batch()
            // userList.weapon.removeAt(userList.weapon.indexOf(Weapon(position.toLong())))
            userList.weapon?.removeAll { it.weaponId == position.toLong() }
            batch.update(userDocument, "weapon", userList.weapon)
            //    when {
//                isInInventory() -> {
//                    val weapon = armingList.weapon
//                    weapon.removeAt(weapon.indexOf(Weapon(position.toLong())))
//                    batch.update(armingDocumentReference!!, "weapon", weapon)
//                }
//                isInstalled() -> {
//                    val weapon = userList.weapon
//                    weapon.removeAt(weapon.indexOf(Weapon(position.toLong())))
//                    batch.update(userDocumentReference!!, "weapon", weapon)
//                }
//                else -> Snackbar.make(
//                        modulesinfo_parentLayout,
//                        "The item has not been found",
//                        Snackbar.LENGTH_SHORT)
//                        .show() //   }

//            val moduleMap = HashMap<String, Any>()
//            moduleMap["damageHP"] = Utils.getCurrentModuleInfo(0L)!!.damageHP
//            moduleMap["weaponClass"] = Utils.getCurrentModuleInfo(0L)!!.moduleClass
//            moduleMap["weaponName"] = Utils.getCurrentModuleInfo(0L)!!.name
            batch.update(userDocument, "money", userList.money!! + module?.price!! / 2)
            batch.commit()
                    .addOnCompleteListener {
                        if (isEnoughMoneyToBuy()) {
                            binding.modulesinfoBuy.isEnabled = true
                            binding.modulesinfoBuy.setBackgroundColor(
                                    ContextCompat.getColor(this, R.color.colorAccent)) }
                        if (isInInventory()) {
                            binding.modulesinfoSell.isEnabled = true
                            binding.modulesinfoSell.setBackgroundColor(ContextCompat.getColor(this, R.color.colorAccent))
                        }
                        showSnackbarError(getString(R.string.modules_info_item_was_sold_successfully))
                    }
        }
    }

    override fun onStart() {
        super.onStart()
        userDocument.addSnapshotListener(
                this
        ) { doc, _ ->
            if (doc!!.exists()) {
                userList = doc.toObject(User::class.java)!!
                binding.modulesinfoUserCash.text = userList.money.toString()
                setOnSellButtonVisibility(isInstalled())
                setOnBuyButtonVisibility(isEnoughMoneyToBuy() && !isInstalled())
            }
        }
    }

    private fun isInstalled(): Boolean {
        return userList.weapon!!.any { it.weaponId == position.toLong() }
    }

    private fun isInInventory(): Boolean {
        return userList.weapon!!.contains(Weapon(position.toLong()))
    }

    private fun isEnoughMoneyToBuy(): Boolean {
        return userList.money!! >= module?.price!!
    }

    private fun isSlotAvailable(): Boolean {
        return userList.weapon?.filter { it.isWeaponInstalled == true }?.size!! < getCurrentShipInfo(userList.ship!!).weaponSlots!!
    }

    private fun setOnBuyButtonVisibility(state: Boolean) {
        binding.modulesinfoBuy.isEnabled = state
        if (state) {
            binding.modulesinfoBuy.setBackgroundColor(ContextCompat.getColor(this, R.color.colorAccent))
        } else {
            binding.modulesinfoBuy.setBackgroundColor(ContextCompat.getColor(this, R.color.grey))
        }
    }

    private fun setOnSellButtonVisibility(state: Boolean) {
        binding.modulesinfoSell.isEnabled = state
        if (state) {
            binding.modulesinfoSell.setBackgroundColor(ContextCompat.getColor(this, R.color.colorAccent))
        } else {
            binding.modulesinfoSell.setBackgroundColor(ContextCompat.getColor(this, R.color.grey))
        }
    }

    fun onGoBackToMainOptions(view: View) {
        finish()
    }
}
